### ISSUE TITLE
XGBoost Spark and PMML score mismatch

### ISSUE THREAD

**Issue by srivatsan88** [jpmml/jpmml-xgboost#27]
Hi,

I have a model trained in Spark using below method
```
val xgboostModel = XGBoost.trainWithDataFrame(<spark df>,...) . Once trained we are converting it to booster model
val modelbooster=xgboostModel.booster
modelbooster.saveModel("<location>")
```
We are using jpmml-xgboost to to convert to pmml and jpmml evaluator (version 1.3.4) to score the same.

I am noticing the score produced by native spark transform method to that of pmml is not matching. While it matched for 60% of the records there is a 40% mismatch.

Note: We are currently not using jpmml-spark-xgboost as we are not using pipeline today. Is there a way we can generate pmml using XGBoost model rather XGB Estimator in spark

Another point is when we score the booster model in python surprisingly it matched to pmml evaluator score. Is there an issue with Spark conversion to Booster that you have noticed and is there a work around for the same?

**Comment by vruusmann** [comment]
The JPMML-XGBoost library operates on the binary dump file. It doesn't matter if the dump is generated using R, Scikit-Learn, Apache Spark or standalone XGBoost - the input to the converter, and the conversion logic itself, is always the same.

The problem is that the Apache Spark integration of XGBoost is generating bad dump files. It's somehow related to the encoding of categorical features - if your model uses only continuous and binary indicator features, then the dump file is OK, but if there is at least one categorical feature involved, then the dump file is bad.

In other words, this is a XGBoost4J bug, not a JPMML-XGBoost (or JPMML-Evaluator) bug. I don't have time to debug this (as I'm unable to build XGBoost4J from source checkout on my platform), but if anyone else is interested, then that would be interesting to see.

Moreover, if you load Apache Spark generated XGBoost dump file in R or Scikit-Learn code, then you would be getting mismatching predictions as well.

**Comment by vruusmann** [comment]
> Will reach out to XGBoost community and check it out as well.

@srivatsan88 The problem could kick in very early, perhaps the training `DMatrix` is prepared incorrectly when there are categorical features involved? And a bad DMatrix leads to a bad XGBoost model, which leads to a bad dump file.

Back in the day, I was training an Apache Spark XGBoost model with a toy dataset, which contained only one categorical feature (at least three categories deep, as binary categorical features seem to be OK).
